### PR TITLE
Fix #3034, #3032 - Bookmarks performance & adding/editing & displaying last selected folder/default folder

### LINKED ISSUE #3032: Crash when trying to add or edit bookmark
### Description: 

Crash when trying to add or edit bookmark on Apple iPad Air 2, iPadOS 14.2

### Steps to Reproduce 

1) Existing sync chain with Windows 10 PC and Android phone. (Don’t know if both devices are needed)
2) Install brave 1.21 as new installation.
3) Add iPad to the sync chain.
- Try adding a bookmark on the iPad. Crash
- Try editing existing bookmark on the iPad. Crash.

Removing sync from the iPad makes no change.
Removing sync from the iPad and deleting all bookmarks makes (edit remove: above work.)
(edit add: it possible to add a bookmark and edit it)

**Actual result:** 
Crash

**Expected result:**
Be able to add or edit bookmark.

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easy (for me)

**Brave Version:** 

**Device details:** 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### LINKED ISSUE #3034: URL autocompletion doesn't suggest bookmarks anymore
Regression caused by moving over to the new sync implementation, and having to use new bookmarks model

### LINKED ISSUE #3045: Open bookmarks screen at last visited folder
Users are complaining about having to use 2 taps to go to `Mobile bookmarks` folder,
with this enhancement we will always open it at level of last visited folder
if not possible to visit the folder we fallback to `Mobile Bookmarks` folder
if `Mobile Bookmarks` is not available we fallback to root folder.

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary of Changes
- Fixes DSYMs not being found when using BraveCore with node modules (local builds only).
- Fixes performance issues found when a LARGE amount of bookmarks are added, synced, and the user goes to add more or edit existing ones.
- Fixes performance issues with displaying of bookmark icons when a large amount of bookmarks are loaded.
- Fixes search suggestions regression so that when the user searches in the URL bar, it will suggest bookmarks as well.
- Adds ability to display the last viewed folder by default. If no folder is viewed, it displays the mobile bookmarks folder. If the root folder was viewed, it displays that. IE: It will always display the last viewed folder by default, and fallback to the mobile bookmarks folder.

With these changes, bookmarks are more stabled and way more performant!

This pull request fixes: #3034, #3032, #3045 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

As this PR fixes few tickets at once, test plans are added to each ticket separately
#3034 
#3032 
#3045 

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarksViewController.swift [hunk 1]
```diff
@@ -587,3 +662,16 @@ extension BookmarksViewController {
         }
     }
 }
+
+extension UIViewController {
```
Comment (iccub) [inline Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarksViewController.swift]: Can we make this `private`?

**Review by iccub — CHANGES_REQUESTED** [review]
code looks good, need updated brave-core framework to fully test it

File: Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarksViewController.swift (hunk 1 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/BookmarksViewController.swift]: Fixed.

File: Client/Frontend/Browser/Search/SearchLoader.swift [hunk 2]
```diff
@@ -63,7 +63,9 @@ class _SearchLoader<UnusedA, UnusedB>: Loader<[Site], SearchViewController> {
                 }
                 
                 // First, see if the query matches any URLs from the user's search history.
-                self.load(result)
+                let bookmarks = Set<Site>(Bookmarkv2.byFrequency(query: self.query).map { Site(url: $0.url ?? "", title: $0.title ?? "", bookmarked: true) })
```
Comment (iccub) [inline Client/Frontend/Browser/Search/SearchLoader.swift:66]: This should show bookmarks only - exclude folders.

**Review by iccub — CHANGES_REQUESTED** [review]

**Review by iccub — APPROVED** [review]

**Review by iccub — APPROVED** [review]

**Pull request merged**